File: src/logic/watch-command-processor.ts
import type { AppSettings, CurrentState, Turno } from "../shared/types";
import type { WatchCommand, WatchCommandResponse, WatchEntry, WatchEntryType, WatchTurno, WatchTurnoTotals } from "../shared/watch-commands";
import { buildTurnoConfigFromSettings, calcularTurnoContable } from "./accounting";
import { fmtDuration } from "./formatters";
import { mergeTurnos, sortTurnosByDateDesc } from "./turnos";

export type WatchCommandProcessorState = {
  current: CurrentState;
  history: Turno[];
  processedOperationIds: string[];
  settings: Pick<
    AppSettings,
    | "porcentaje.jefe"
    | "porcentaje.chofer"
    | "descontar.datafono"
    | "descontar.agencia_bono"
    | "descontar.extra"
    | "descontar.gasolina"
    | "diaLibre"
  >;
  now: {
    date: string;
    time: string;
    id: number;
  };
};

export type WatchCommandProcessorResult = WatchCommandProcessorState & {
  response: WatchCommandResponse;
};

function isActive(current: CurrentState): boolean {
  return current.entries.length > 0 || !!current.startTime;
}

/** Suma importes y recuenta entradas del turno en curso, por categoría. */
export function computeWatchTotals(current: CurrentState): WatchTurnoTotals {
  const porTipo: Record<WatchEntryType, number> = {
    propina: 0,
    datafono: 0,
    agencia_bono: 0,
    extra: 0,
    gasolina: 0,
    nulo: 0,
  };
  const numPorTipo: Record<WatchEntryType, number> = {
    propina: 0,
    datafono: 0,
    agencia_bono: 0,
    extra: 0,
    gasolina: 0,
    nulo: 0,
  };
  for (const entry of current.entries) {
    if (entry.type in porTipo) {
      porTipo[entry.type as WatchEntryType] += entry.amount;
      numPorTipo[entry.type as WatchEntryType] += 1;
    }
  }
  return { porTipo, numPorTipo, numEntradas: current.entries.length };
}

/** Historial del turno para el reloj: más recientes primero. */
export function buildWatchEntradas(current: CurrentState): WatchEntry[] {
  return current.entries
    .map((e) => ({
      id: e.id,
      type: e.type as WatchEntry["type"],
      amount: e.amount,
      note: e.note,
      time: e.time,
    }))
    .reverse();
}

function buildWatchTotalsFromTurno(turno: Turno): WatchTurnoTotals {
  const porTipo: Record<WatchEntryType, number> = {
    propina: turno.totalP || 0,
    datafono: turno.totalD || 0,
    agencia_bono: turno.totalA || 0,
    extra: turno.totalE || 0,
    gasolina: turno.totalF || 0,
    nulo: turno.totalN || 0,
  };
  const numPorTipo: Record<WatchEntryType, number> = {
    propina: 0,
    datafono: 0,
    agencia_bono: 0,
    extra: 0,
    gasolina: 0,
    nulo: 0,
  };

  for (const entry of turno.entries) {
    if (entry.type in numPorTipo) {
      numPorTipo[entry.type as WatchEntryType] += 1;
    }
  }

  return { porTipo, numPorTipo, numEntradas: turno.entries.length };
}

function buildWatchEntradasFromTurno(turno: Turno): WatchEntry[] {
  return turno.entries
    .map((e) => ({
      id: e.id,
      type: e.type as WatchEntry["type"],
      amount: e.amount,
      note: e.note,
      time: e.time,
    }))
    .reverse();
}

export function buildWatchTurnos(history: Turno[], settings: WatchCommandProcessorState["settings"]): WatchTurno[] {
  return sortTurnosByDateDesc(history).slice(0, 30).map((turno) => {
    let totalMins = 0;
    if (turno.startTime && turno.endTime) {
      const [startH, startM] = turno.startTime.split(":").map(Number);
      const [endH, endM] = turno.endTime.split(":").map(Number);
      if (Number.isFinite(startH) && Number.isFinite(startM) && Number.isFinite(endH) && Number.isFinite(endM)) {
        totalMins = (endH * 60 + endM) - (startH * 60 + startM);
        if (totalMins < 0) totalMins += 24 * 60;
        totalMins = Math.max(0, totalMins - (turno.totalPausedMinutes || 0));
      }
    }
    const calculo = calcularTurnoContable(turno, settings);

    return {
      id: turno.id,
      date: turno.date,
      startDate: turno.startDate,
      startTime: turno.startTime,
      endTime: turno.endTime,
      dinero: turno.dinero || 0,
      km: turno.km || 0,
      totalTaximetro: calculo.dineroBase,
      miGanancia: calculo.miGanancia,
      totalADescontar: calculo.totalDescontar,
      totalADar: calculo.totalADar,
      tiempoTrabajado: fmtDuration(totalMins),
      totalPausedMinutes: turno.totalPausedMinutes || 0,
      totals: buildWatchTotalsFromTurno(turno),
      entradas: buildWatchEntradasFromTurno(turno),
    };
  });
}

function emptyCurrent(): CurrentState {
  return {
    entries: [],
    startTime: null,
    startDate: null,
    isPaused: false,
    pauseStartTime: null,
    totalPausedMinutes: 0,
  };
}

function withProcessedOperationId(
  state: WatchCommandProcessorState,
  operationId: string,
): string[] {
  return state.processedOperationIds.includes(operationId)
    ? state.processedOperationIds
    : [...state.processedOperationIds, operationId];
}

function errorResponse(command: WatchCommand, code: string, message: string): WatchCommandResponse {
  return {
    type: "ERROR",
    operationId: command.operationId,
    code,
    message,
  };
}

function elapsedMinutes(startTime: string | null | undefined, endTime: string): number {
  if (!startTime) return 0;
  const [startHour, startMinute] = startTime.split(":").map(Number);
  const [endHour, endMinute] = endTime.split(":").map(Number);
  if (![startHour, startMinute, endHour, endMinute].every(Number.isFinite)) return 0;
  let elapsed = (endHour * 60 + endMinute) - (startHour * 60 + startMinute);
  if (elapsed < 0) elapsed += 24 * 60;
  return elapsed;
}

export function processWatchCommand(
  command: WatchCommand,
  state: WatchCommandProcessorState,
): WatchCommandProcessorResult {
  if (!command.operationId.trim()) {
    return {
      ...state,
      response: errorResponse(command, "INVALID_OPERATION_ID", "operationId obligatorio"),
    };
  }

  if (command.type === "GET_STATUS") {
    return {
      ...state,
      response: {
        type: "STATUS",
        connected: true,
        activeTurno: isActive(state.current),
        startTime: state.current.startTime,
        startDate: state.current.startDate,
        totals: computeWatchTotals(state.current),
        entradas: buildWatchEntradas(state.current),
        isPaused: state.current.isPaused ?? false,
        pauseStartTime: state.current.pauseStartTime ?? null,
        totalPausedMinutes: state.current.totalPausedMinutes ?? 0,
      },
    };
  }

  if (command.type === "GET_TURNOS") {
    return {
      ...state,
      response: {
        type: "TURNOS_STATUS",
        connected: true,
        turnos: buildWatchTurnos(state.history, state.settings),
      },
    };
  }

  if (state.processedOperationIds.includes(command.operationId)) {
    return {
      ...state,
      response: {
        type: "DUPLICATE_IGNORED",
        operationId: command.operationId,
        message: "Operacion ya procesada",
      },
    };
  }

  if (command.type === "START_TURNO") {
    if (isActive(state.current)) {
      return {
        ...state,
        response: errorResponse(command, "ACTIVE_TURNO", "Ya hay turno activo"),
      };
    }

    return {
      ...state,
      current: {
        ...state.current,
        startTime: state.now.time,
        startDate: state.now.date,
      },
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Turno iniciado",
      },
    };
  }

  if (command.type === "PAUSE_TURNO") {
    if (!state.current.startTime) {
      return {
        ...state,
        response: errorResponse(command, "NO_ACTIVE_TURNO", "No hay turno activo"),
      };
    }
    if (state.current.isPaused) {
      return {
        ...state,
        response: errorResponse(command, "ALREADY_PAUSED", "El turno ya esta pausado"),
      };
    }

    return {
      ...state,
      current: {
        ...state.current,
        isPaused: true,
        pauseStartTime: state.now.time,
      },
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Turno pausado",
      },
    };
  }

  if (command.type === "RESUME_TURNO") {
    if (!state.current.startTime) {
      return {
        ...state,
        response: errorResponse(command, "NO_ACTIVE_TURNO", "No hay turno activo"),
      };
    }
    if (!state.current.isPaused || !state.current.pauseStartTime) {
      return {
        ...state,
        response: errorResponse(command, "NOT_PAUSED", "El turno no esta pausado"),
      };
    }

    return {
      ...state,
      current: {
        ...state.current,
        isPaused: false,
        pauseStartTime: null,
        totalPausedMinutes:
          (state.current.totalPausedMinutes || 0) +
          elapsedMinutes(state.current.pauseStartTime, state.now.time),
      },
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Turno reanudado",
      },
    };
  }

  if (command.type === "ADD_ENTRY") {
    if (!state.current.startTime) {
      return {
        ...state,
        response: errorResponse(command, "NO_ACTIVE_TURNO", "No hay turno activo"),
      };
    }
    if (!(command.payload.amount > 0)) {
      return {
        ...state,
        response: errorResponse(command, "INVALID_AMOUNT", "Importe invalido"),
      };
    }

    return {
      ...state,
      current: {
        ...state.current,
        entries: [
          ...state.current.entries,
          {
            id: state.now.id,
            type: command.payload.entryType,
            amount: command.payload.amount,
            note: command.payload.note.trim(),
            time: state.now.time,
          },
        ],
      },
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Entrada anadida",
      },
    };
  }

  if (command.type === "ADD_NOTE") {
    if (!state.current.startTime) {
      return {
        ...state,
        response: errorResponse(command, "NO_ACTIVE_TURNO", "No hay turno activo"),
      };
    }
    const note = command.payload.note.trim();
    if (!note) {
      return {
        ...state,
        response: errorResponse(command, "INVALID_NOTE", "Nota obligatoria"),
      };
    }

    return {
      ...state,
      current: {
        ...state.current,
        entries: [
          ...state.current.entries,
          {
            id: state.now.id,
            type: "nota",
            amount: 0,
            note,
            time: state.now.time,
          },
        ],
      },
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Nota anadida",
      },
    };
  }

  if (command.type === "EDIT_ENTRY") {
    if (!state.current.startTime) {
      return {
        ...state,
        response: errorResponse(command, "NO_ACTIVE_TURNO", "No hay turno activo"),
      };
    }
    const target = state.current.entries.find((e) => e.id === command.payload.id);
    if (!target) {
      return {
        ...state,
        response: errorResponse(command, "ENTRY_NOT_FOUND", "Entrada no encontrada"),
      };
    }
    if (target.type !== "nota" && !(command.payload.amount > 0)) {
      return {
        ...state,
        response: errorResponse(command, "INVALID_AMOUNT", "Importe invalido"),
      };
    }

    return {
      ...state,
      current: {
        ...state.current,
        entries: state.current.entries.map((e) =>
          e.id === command.payload.id
            ? { ...e, amount: e.type === "nota" ? e.amount : command.payload.amount, note: command.payload.note.trim() }
            : e,
        ),
      },
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Entrada editada",
      },
    };
  }

  if (command.type === "DELETE_ENTRY") {
    if (!state.current.startTime) {
      return {
        ...state,
        response: errorResponse(command, "NO_ACTIVE_TURNO", "No hay turno activo"),
      };
    }
    const exists = state.current.entries.some((e) => e.id === command.payload.id);
    if (!exists) {
      return {
        ...state,
        response: errorResponse(command, "ENTRY_NOT_FOUND", "Entrada no encontrada"),
      };
    }

    return {
      ...state,
      current: {
        ...state.current,
        entries: state.current.entries.filter((e) => e.id !== command.payload.id),
      },
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Entrada borrada",
      },
    };
  }

  if (command.type === "END_TURNO") {
    if (!state.current.startTime) {
      return {
        ...state,
        response: errorResponse(command, "NO_ACTIVE_TURNO", "No hay turno activo"),
      };
    }
    if (!(command.payload.dinero > 0) || !(command.payload.km > 0)) {
      return {
        ...state,
        response: errorResponse(command, "INVALID_END_VALUES", "Taximetro y kilometros obligatorios"),
      };
    }

    const entries = state.current.entries;
    const turno: Turno = {
      id: state.now.id,
      date: state.now.date,
      startTime: state.current.startTime,
      endTime: state.now.time,
      entries,
      totalP: entries.filter((e) => e.type === "propina").reduce((s, e) => s + e.amount, 0),
      totalD: entries.filter((e) => e.type === "datafono").reduce((s, e) => s + e.amount, 0),
      totalA: entries.filter((e) => e.type === "agencia_bono").reduce((s, e) => s + e.amount, 0),
      totalE: entries.filter((e) => e.type === "extra").reduce((s, e) => s + e.amount, 0),
      totalF: entries.filter((e) => e.type === "gasolina").reduce((s, e) => s + e.amount, 0),
      totalN: entries.filter((e) => e.type === "nulo").reduce((s, e) => s + e.amount, 0),
      dinero: command.payload.dinero,
      km: command.payload.km,
      notes: command.payload.note.trim(),
      startDate: state.current.startDate,
      totalPausedMinutes:
        (state.current.totalPausedMinutes || 0) +
        (state.current.isPaused ? elapsedMinutes(state.current.pauseStartTime, state.now.time) : 0),
      configTurno: buildTurnoConfigFromSettings(state.settings),
      diaLibreContable: state.settings.diaLibre,
    };

    return {
      ...state,
      current: emptyCurrent(),
      history: mergeTurnos(state.history, [turno]),
      processedOperationIds: withProcessedOperationId(state, command.operationId),
      response: {
        type: "OK",
        operationId: command.operationId,
        message: "Turno terminado",
      },
    };
  }

  return {
    ...state,
    response: errorResponse(command, "UNKNOWN_COMMAND", "Comando no reconocido"),
  };
}

File: android/app/src/main/java/com/mijornada/app/watch/WatchCommandProcessor.kt
package com.mijornada.app.watch

object WatchCommandProcessor {
    @JvmStatic
    fun process(command: WatchCommand, state: WatchProcessorState, operationExists: (String) -> Boolean): WatchProcessorResult {
        if (command.operationId.isBlank()) {
            return state.error(command.operationId, "INVALID_OPERATION_ID", "operationId obligatorio")
        }

        if (operationExists(command.operationId)) {
            return WatchProcessorResult(
                state = state,
                response = WatchResponse.DuplicateIgnored(command.operationId, "Operacion ya procesada"),
            )
        }

        return when (command) {
            is WatchCommand.StartTurno -> processStartTurno(command, state)
            is WatchCommand.PauseTurno -> processPauseTurno(command, state)
            is WatchCommand.ResumeTurno -> processResumeTurno(command, state)
            is WatchCommand.AddEntry -> processAddEntry(command, state)
            is WatchCommand.AddNote -> processAddNote(command, state)
            is WatchCommand.EditEntry -> processEditEntry(command, state)
            is WatchCommand.DeleteEntry -> processDeleteEntry(command, state)
            is WatchCommand.EndTurno -> processEndTurno(command, state)
        }
    }

    private fun processPauseTurno(command: WatchCommand.PauseTurno, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.startTime == null) {
            return state.error(command.operationId, "NO_ACTIVE_TURNO", "No hay turno activo")
        }
        if (state.current.isPaused) {
            return state.error(command.operationId, "ALREADY_PAUSED", "El turno ya esta pausado")
        }
        return WatchProcessorResult(
            state = state.copy(
                current = state.current.copy(isPaused = true, pauseStartTime = state.nowTime),
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Turno pausado"),
        )
    }

    private fun processResumeTurno(command: WatchCommand.ResumeTurno, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.startTime == null) {
            return state.error(command.operationId, "NO_ACTIVE_TURNO", "No hay turno activo")
        }
        if (!state.current.isPaused || state.current.pauseStartTime == null) {
            return state.error(command.operationId, "NOT_PAUSED", "El turno no esta pausado")
        }
        val pausedMinutes = elapsedMinutes(state.current.pauseStartTime, state.nowTime)
        return WatchProcessorResult(
            state = state.copy(
                current = state.current.copy(
                    isPaused = false,
                    pauseStartTime = null,
                    totalPausedMinutes = state.current.totalPausedMinutes + pausedMinutes,
                ),
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Turno reanudado"),
        )
    }

    private fun processStartTurno(command: WatchCommand.StartTurno, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.isActive()) {
            return state.error(command.operationId, "ACTIVE_TURNO", "Ya hay turno activo")
        }

        return WatchProcessorResult(
            state = state.copy(
                current = state.current.copy(startTime = state.nowTime, startDate = state.nowDate),
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Turno iniciado"),
        )
    }

    private fun processAddEntry(command: WatchCommand.AddEntry, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.startTime == null) {
            return state.error(command.operationId, "NO_ACTIVE_TURNO", "No hay turno activo")
        }
        if (command.amount <= 0.0) {
            return state.error(command.operationId, "INVALID_AMOUNT", "Importe invalido")
        }

        val entry = WatchEntry(
            id = state.nowId,
            type = command.entryType,
            amount = command.amount,
            note = command.note.trim(),
            time = state.nowTime,
        )

        return WatchProcessorResult(
            state = state.copy(
                current = state.current.copy(entries = state.current.entries + entry),
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Entrada anadida"),
        )
    }

    private fun processAddNote(command: WatchCommand.AddNote, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.startTime == null) {
            return state.error(command.operationId, "NO_ACTIVE_TURNO", "No hay turno activo")
        }
        val note = command.note.trim()
        if (note.isBlank()) {
            return state.error(command.operationId, "INVALID_NOTE", "Nota obligatoria")
        }

        val entry = WatchEntry(
            id = state.nowId,
            type = "nota",
            amount = 0.0,
            note = note,
            time = state.nowTime,
        )

        return WatchProcessorResult(
            state = state.copy(
                current = state.current.copy(entries = state.current.entries + entry),
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Nota anadida"),
        )
    }

    private fun processEditEntry(command: WatchCommand.EditEntry, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.startTime == null) {
            return state.error(command.operationId, "NO_ACTIVE_TURNO", "No hay turno activo")
        }

        val target = state.current.entries.find { it.id == command.id }
            ?: return state.error(command.operationId, "ENTRY_NOT_FOUND", "Entrada no encontrada")

        if (target.type != "nota" && command.amount <= 0.0) {
            return state.error(command.operationId, "INVALID_AMOUNT", "Importe invalido")
        }

        val nextEntries = state.current.entries.map { entry ->
            if (entry.id == command.id) {
                entry.copy(
                    amount = if (entry.type == "nota") entry.amount else command.amount,
                    note = command.note.trim(),
                )
            } else {
                entry
            }
        }

        return WatchProcessorResult(
            state = state.copy(
                current = state.current.copy(entries = nextEntries),
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Entrada editada"),
        )
    }

    private fun processDeleteEntry(command: WatchCommand.DeleteEntry, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.startTime == null) {
            return state.error(command.operationId, "NO_ACTIVE_TURNO", "No hay turno activo")
        }
        if (state.current.entries.none { it.id == command.id }) {
            return state.error(command.operationId, "ENTRY_NOT_FOUND", "Entrada no encontrada")
        }

        return WatchProcessorResult(
            state = state.copy(
                current = state.current.copy(entries = state.current.entries.filter { it.id != command.id }),
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Entrada borrada"),
        )
    }

    private fun processEndTurno(command: WatchCommand.EndTurno, state: WatchProcessorState): WatchProcessorResult {
        if (state.current.startTime == null) {
            return state.error(command.operationId, "NO_ACTIVE_TURNO", "No hay turno activo")
        }
        if (command.dinero <= 0.0 || command.km <= 0.0) {
            return state.error(command.operationId, "INVALID_END_VALUES", "Taximetro y kilometros obligatorios")
        }

        val pendingPausedMinutes = if (state.current.isPaused && state.current.pauseStartTime != null) {
            elapsedMinutes(state.current.pauseStartTime, state.nowTime)
        } else {
            0
        }
        val turno = WatchTurno(
            id = state.nowId,
            date = state.nowDate,
            startDate = state.current.startDate,
            startTime = state.current.startTime,
            endTime = state.nowTime,
            entries = state.current.entries,
            dinero = command.dinero,
            km = command.km,
            notes = command.note.trim(),
            totalPausedMinutes = state.current.totalPausedMinutes + pendingPausedMinutes,
        )

        return WatchProcessorResult(
            state = state.copy(
                current = WatchCurrentState.empty(),
                history = state.history + turno,
                processedOperationIds = state.withOperationId(command.operationId),
            ),
            response = WatchResponse.Ok(command.operationId, "Turno terminado"),
        )
    }

    private fun WatchProcessorState.withOperationId(operationId: String): List<String> =
        if (processedOperationIds.contains(operationId)) processedOperationIds else processedOperationIds + operationId

    private fun elapsedMinutes(startTime: String, endTime: String): Int {
        fun toMinutes(value: String): Int? {
            val parts = value.split(":")
            if (parts.size < 2) return null
            val hours = parts[0].toIntOrNull() ?: return null
            val minutes = parts[1].toIntOrNull() ?: return null
            return hours * 60 + minutes
        }
        val start = toMinutes(startTime) ?: return 0
        val end = toMinutes(endTime) ?: return 0
        val difference = end - start
        return if (difference >= 0) difference else difference + 24 * 60
    }

    private fun WatchProcessorState.error(operationId: String, code: String, message: String): WatchProcessorResult =
        WatchProcessorResult(
            state = this,
            response = WatchResponse.Error(operationId, code, message),
        )
}
